enhancements/separating-provider-specific-code-in-the-installer.md

```diff
@@ -0,0 +1,147 @@
+---
+title: separating-provider-specific-code-in-the-installer
+authors:
+  - "@janoszen"
+reviewers:
+  - "@Gal-Zaidman"
+approvers:
+  - TBD
+creation-date: 2020-11-27
+last-updated: 2020-11-27
+status: provisional
+see-also: []
+replaces: []
+superseded-by: []
+---
+
+# Separating provider-specific code in the installer
+
+## Release Signoff Checklist
+
+TBD
+
+- [ ] Enhancement is `implementable`
+- [ ] Design details are appropriately documented from clear requirements
+- [ ] Test plan is defined
+- [ ] Graduation criteria for dev preview, tech preview, GA
+
+## Summary
+
+This change introduces an internal API in the installer code to move provider-specific code into a separate folder and avoid having large switch-case structures in the installer code.
+
+## Motivation
+
+Several parts of the installer code have large, [unwieldy switch-case structures](https://github.com/openshift/installer/blob/254680f6268a7b8e57091d788256c67b81755b5d/pkg/asset/cluster/tfvars.go#L166) that contain provider-specific codes. This makes it hard to keep track of the code that serves a provider, and it makes it impossible to write component-level tests. Creating a separated API reduces friction between teams and makes changes easier to merge because the risk of accidental side effects between providers is reduced. 
+
+### Goals
+
+Create a Go interface and the hooks in the installer code that separatees the installer core code from the provider-specific code.
+
+### Non-Goals
+
+Remove all provider-specific pieces from the code. That would require a complete refactor of the install config data structure and move the Terraform code. 
+
+## Proposal
+
+### Step 1: Create a Go interface
+
+We should create a Go interface that describes the functionality each provider must implement. For example, from the prototype:
+
+```go
+type PlatformV2 interface {
+	// Name returns the name of the platform.
+	Name() string
+
+	// Metadata translates the installConfig to the appropriate entry in clusterMetadata.
+	Metadata(clusterMetadata *types.ClusterMetadata, installConfig *installconfig.InstallConfig) error
+
+    //...
+}
+```
+
+Not all platforms implement all features. Optional features should be hidden behind a feature flag and a separate interface:
+
+```go
+var ErrNotAnIPIPlatform = errors.New("...")
+
+type PlatformV2 interface {
+    //...
+
+    // GetIPI returns the IPI-specific interface, or ErrNotAnIPIPlatform if the platform does not support IPI.
+	GetIPI() (IPIPlatformV2, error)
+}
+
+type IPIPlatformV2 interface {
+    // IPI-specific 
+}
+``` 
+
+### Step 2: Implement registry
+
+In this step we should implement a registry pattern where providers can register with their names.
```

Comment:
In the OpenStack provider, we have started the effort of defining a clean provider interface for Installer steps. For example, [for `cloudproviderconfig`](
https://github.com/openshift/installer/blob/ec982ecf5bd847add0f3af82b04f32251972701c/pkg/asset/manifests/cloudproviderconfig.go#L97-L105) we have come up with this signature:

```Go
func GenerateCloudProviderConfig(installConfig types.InstallConfig) (cloudProviderConfigData, cloudProviderConfigCABundleData string, err error)
```

We would be happy to be guinea pigs for this RFE and help defining and testing interface signatures in-vivo for the other steps as well! We also have that [in our backlog](https://issues.redhat.com/browse/OSASINFRA-1901).

Reply:
There ought to be a saying about two guinea pigs being better than one, we (oVirt) are also wanting to have this. Let's do it!